test/worker_test.py

```diff
@@ -2135,3 +2136,61 @@ def run(self):
                 self.assertEqual(0, self.sch._state.get_task(s1.task_id).failures.num_failures())
                 self.assertEqual(self.per_task_retry_count, self.sch._state.get_task(e2.task_id).failures.num_failures())
                 self.assertEqual(self.default_retry_count, self.sch._state.get_task(e1.task_id).failures.num_failures())
+
+
+class SleepyLocalTarget(luigi.LocalTarget):
+    time_to_sleep = 0.5
+
+    def exists(self):
+        time.sleep(self.time_to_sleep)
+        return super(luigi.LocalTarget, self).exists()
+
+    def touch(self):
+        if os.path.exists(self.path):
+            raise Exception("Path {} already exists".format(self.path))
+        else:
+            self.open("w").close()
+
+
+class Child(luigi.Task):
+    time_to_sleep = 1
+    output_dir = luigi.Parameter(default="")
+
+    def output(self):
+        return SleepyLocalTarget("{}/Child".format(self.output_dir))
+
+    def run(self):
+        time.sleep(self.time_to_sleep)
+        self.output().touch()
+
+
+class Parent(luigi.Task):
+    time_to_sleep = 2
+    output_dir = luigi.Parameter(default="")
+
+    def requires(self):
+        return Child(output_dir=self.output_dir)
+
+    def output(self):
+        return SleepyLocalTarget("{}/Parent".format(self.output_dir))
+
+    def run(self):
+        time.sleep(self.time_to_sleep)
+        self.output().touch()
+
+
+class ConcurrentTest(unittest.TestCase):
+
+    def setUp(self):
+        self.p = tempfile.mkdtemp()
+
+    def tearDown(self):
+        shutil.rmtree(self.p)
+
+    def test_sending_same_task_twice_does_not_lead_to_double_run(self):
+        process = Process(target=luigi.build,
+                          kwargs={"tasks": [Parent(output_dir=self.p)], "local_scheduler": True})
+        process.start()
+        time.sleep(SleepyLocalTarget.time_to_sleep * 2 + Parent.time_to_sleep + Child.time_to_sleep)
+        self.assertTrue(luigi.build([Parent(output_dir=self.p)], local_scheduler=True))
```

Comment:
I have little idea what this assertTrue shows. Did this fail before?? Did it fail because nothing was executed or what?

There is a (very) new thing called LuigiRunResult which you can use here to make it clearer what the test case is checking (it also adds an enum that could improve clarity further)

Reply:
Let me clarify, there's a `SleepyLocalTarget` that does not allow to run the task twice: it fails with an exception. So if the `Parent` task is run twice, `luigi.build` will return False. This test fails without the changes in `worker.py` in this PR since luigi tries to run `Parent` task twice.
Do you think this can be expressed using `LuigiRunResult` in a clearer way? If yes, I'll give it a try.